src/librustc/middle/resolve_lifetime.rs

```diff
@@ -638,8 +638,8 @@ impl<'a, 'tcx> Visitor<'tcx> for LifetimeContext<'a, 'tcx> {
                     // and ban them. Type variables instantiated inside binders aren't
                     // well-supported at the moment, so this doesn't work.
                     // In the future, this should be fixed and this error should be removed.
-                    let def = self.map.defs.get(&lifetime.id).cloned();
-                    if let Some(Region::LateBound(_, def_id, _)) = def {
+                    let def = self.map.defs.get(&lifetime.id);
+                    if let Some(&Region::LateBound(_, def_id, _)) = def {
```

Comment:
It still seems like it might be good not to generate the late-bound entries we do, as they are pretty malformed, and can lead to ICEs and confusion down the line.

Reply:
I'll reinstate your removal code